server/src/main/java/org/jboss/as/server/deployment/DeploymentUnitPhaseService.java

```diff
@@ -277,4 +281,67 @@ private static boolean shouldRun(final DeploymentUnit unit, final RegisteredDepl
         }
         return !shouldNotRun.contains(deployer.getSubsystemName());
     }
+
+    static class DeploymentUnitServiceTarget extends DelegatingServiceTarget implements ServiceTarget {
+        private final CapabilityServiceSupport support;
+
+        DeploymentUnitServiceTarget(org.jboss.msc.service.ServiceTarget target, CapabilityServiceSupport support) {
+            super(target);
+            this.support = support;
+        }
+
+        @Override
+        public ServiceTarget addListener(LifecycleListener listener) {
+            super.addListener(listener);
+            return this;
+        }
+
+        @Override
+        public ServiceTarget removeListener(LifecycleListener listener) {
+            super.removeListener(listener);
+            return this;
+        }
+
+        @Override
+        public ServiceTarget subTarget() {
+            return new DeploymentUnitServiceTarget(super.subTarget(), this.support);
+        }
+
+        @Override
+        public ServiceBuilder<?> addService() {
+            return new DeploymentUnitServiceBuilder<>(super.addService(), this.support);
+        }
+    }
+
+    static class DeploymentUnitServiceBuilder<T> extends DelegatingServiceBuilder<T> implements ServiceBuilder<T> {
+        private final CapabilityServiceSupport support;
+
+        DeploymentUnitServiceBuilder(org.jboss.msc.service.ServiceBuilder<T> builder, CapabilityServiceSupport support) {
+            super(builder);
+            this.support = support;
+        }
+
+        @Override
+        public ServiceBuilder<T> setInitialMode(Mode mode) {
```

Comment:
Same here. Why returning ServiceBuilder instead of DeploymentUnitServiceBuilder?

Reply:
Because there is nothing to be gained by overloading the return type in the implementation, which is not visible to consumers.

Reply:
Makes sense. Thanks for clarification.